lib/SILOptimizer/LoopTransforms/LICM.cpp

```diff
@@ -395,56 +421,253 @@ void LoopTreeOptimization::propagateSummaries(
+      continue;
+    }
+    return false;
+  }
+  return true;
+}
+
+static void analyzeBeginAccess(BeginAccessInst *BI,
+                               SmallVector<BeginAccessInst *, 8> &BeginAccesses,
+                               SmallVector<EndAccessInst *, 8> &EndAccesses,
+                               HoistPairSet &HoistingPairs) {
+  if (BI->getEnforcement() != SILAccessEnforcement::Dynamic) {
+    return;
+  }
+
+  const AccessedStorage &storage =
+      findAccessedStorageNonNested(BI->getSource());
+  if (!storage) {
+    return;
+  }
+
+  // find matching end accesses:
+  InstSet matchingEnds;
+  auto matchingEndPred = [&](EndAccessInst *EI) {
+    return EI->getBeginAccess() == BI;
+  };
+  for (auto matchingEnd = std::find_if(EndAccesses.begin(), EndAccesses.end(),
+                                       matchingEndPred);
+       matchingEnd != EndAccesses.end();
+       matchingEnd =
+           std::find_if(matchingEnd, EndAccesses.end(), matchingEndPred)) {
+    auto *EI = *matchingEnd;
+    matchingEnds.insert(EI);
+    ++matchingEnd;
+  }
+  if (matchingEnds.empty()) {
+    // no matching end within the loop
+    return;
+  }
+  DEBUG(llvm::dbgs() << "Found " << matchingEnds.size() << " End accesses"
+                     << "\n");
+
+  auto BIAccessedStorageNonNested = findAccessedStorageNonNested(BI);
+  auto safeBeginPred = [&](BeginAccessInst *OtherBI) {
+    if (BI == OtherBI) {
+      return true;
+    }
+    return BIAccessedStorageNonNested.isDistinctFrom(
+        findAccessedStorageNonNested(OtherBI));
+  };
+
+  if (std::all_of(BeginAccesses.begin(), BeginAccesses.end(), safeBeginPred)) {
+    HoistingPairs.emplace_back(std::make_pair(BI, std::move(matchingEnds)));
+  }
+}
+
+// Analyzes current loop for hosting/sinking potential:
+// Computes set of instructions we may be able to move out of the loop
+// Important Note:
+// We can't bail out of this method! we have to run it on all loops.
+// We *need* to discover all MayWrites -
+// even if the loop is otherwise skipped!
+// This is because outer loops will depend on the inner loop's writes.
 void LoopTreeOptimization::analyzeCurrentLoop(
-    std::unique_ptr<LoopNestSummary> &CurrSummary, ReadSet &SafeReads) {
+    std::unique_ptr<LoopNestSummary> &CurrSummary) {
   WriteSet &MayWrites = CurrSummary->MayWrites;
   SILLoop *Loop = CurrSummary->Loop;
   DEBUG(llvm::dbgs() << " Analyzing accesses.\n");
 
   // Contains function calls in the loop, which only read from memory.
   SmallVector<ApplyInst *, 8> ReadOnlyApplies;
+  // Contains Loads inside the loop.
+  SmallVector<LoadInst *, 8> Loads;
+  // Contains fix_lifetime, we might be able to sink them.
+  SmallVector<FixLifetimeInst *, 8> FixLifetimes;
+  // Contains begin_access, we might be able to hoist them.
+  SmallVector<BeginAccessInst *, 8> BeginAccesses;
+  // Contains end_access, we might be able to sink them.
+  SmallVector<EndAccessInst *, 8> EndAccesses;
```

Comment:
This will all be much simpler if you remove this `EndAccesses` set. Then you don't need the `HoistPairSet` or `hoistAndSinkInstructionPair`. In general, we should reason about and handle `begin_access` as a single optimization unit. See the `swift::removeBeginAccess` utility for reference. You can just remove the existing end markers and splat new ones in the loop exit.

Reply:
Hmm good point.. took me a while to think about it, at first I thought pairs can be extended to cover other types of instruction pairs that do not share this property (retain / release for example), but I do like the simplicity of what you suggest. And we can have a simpler transformation for now. I implemented this refactoring as the newest commit in this PR. Pretty straightforward change.

Reply:
That's fine. I was actually suggesting that for any scoped instruction, where all cyclic paths visit a single begin and single end, you just hoist the begin and rewrite the ends. There's no need to call `sinkInstruction`.